File: archive/j/java/MinimumSpanningTree.java
import java.util.Arrays;

public class MinimumSpanningTree {

    public static void main(String[] args) {
        int[] matrix = parse(args);

        int size = (int) Math.sqrt(matrix.length);
        long result = minimumSpanningTree(matrix, size);

        System.out.println(result);
    }

    private static void usage() {
        System.err.println("Usage: please provide a comma-separated list of integers");
        System.exit(1);
    }

    private static int[] parse(String[] args) {
        if (args.length != 1 || args[0] == null || args[0].isBlank()) {
            usage();
        }

        String[] tokens = args[0].split(",");
        int[] values = new int[tokens.length];

        for (int i = 0; i < tokens.length; i++) {
            String token = tokens[i].trim();
            if (token.isEmpty()) {
                usage();
            }

            try {
                values[i] = Integer.parseInt(token);
            } catch (NumberFormatException e) {
                usage();
            }
        }

        int n = (int) Math.sqrt(values.length);
        if (n * n != values.length) {
            usage();
        }

        return values;
    }

    private static long minimumSpanningTree(int[] graph, int size) {
        boolean[] inMst = new boolean[size];
        long[] minEdgeWeight = new long[size];

        Arrays.fill(minEdgeWeight, Long.MAX_VALUE);
        minEdgeWeight[0] = 0;

        long totalWeight = 0;

        for (int i = 0; i < size; i++) {
            int selectedNode = -1;
            long bestWeight = Long.MAX_VALUE;

            for (int v = 0; v < size; v++) {
                if (inMst[v] || minEdgeWeight[v] >= bestWeight) {
                    continue;
                }

                bestWeight = minEdgeWeight[v];
                selectedNode = v;
            }

            if (selectedNode == -1 || bestWeight == Long.MAX_VALUE) {
                usage();
            }

            inMst[selectedNode] = true;
            totalWeight += bestWeight;

            int rowStart = selectedNode * size;

            for (int v = 0; v < size; v++) {
                int edgeWeight = graph[rowStart + v];

                if (inMst[v] || edgeWeight <= 0 || edgeWeight >= minEdgeWeight[v]) {
                    continue;
                }

                minEdgeWeight[v] = edgeWeight;
            }
        }

        return totalWeight;
    }
}

File: archive/j/javascript/minimum-spanning-tree.js
"use strict";

const USAGE = "Usage: please provide a comma-separated list of integers";

const solveMST = (flat, size) => {
  const keys = new Array(size).fill(Infinity);
  const inMST = new Uint8Array(size);

  keys[0] = 0;

  for (let count = 0; count < size; count++) {
    let u = -1;

    for (let v = 0; v < size; v++) {
      if (!inMST[v] && (u === -1 || keys[v] < keys[u])) {
        u = v;
      }
    }

    if (u === -1 || keys[u] === Infinity) break;

    inMST[u] = 1;

    const rowOffset = u * size;
    for (let v = 0; v < size; v++) {
      const weight = flat[rowOffset + v];
      if (weight > 0 && !inMST[v] && weight < keys[v]) {
        keys[v] = weight;
      }
    }
  }

  return keys.reduce((sum, w) => (w === Infinity ? sum : sum + w), 0);
};

const run = () => {
  const [, , input] = process.argv;

  if (!input) {
    console.error(USAGE);
    process.exit(1);
  }

  try {
    const flat = input
      .split(",")
      .filter((x) => x.trim() !== "")
      .map((x) => {
        const val = Number(x);
        if (!Number.isInteger(val)) throw new Error();
        return val;
      });

    const size = Math.sqrt(flat.length);

    if (flat.length === 0 || !Number.isInteger(size)) {
      throw new Error();
    }

    const result = solveMST(flat, size);
    console.log(result);
  } catch {
    console.error(USAGE);
    process.exit(1);
  }
};

run();
